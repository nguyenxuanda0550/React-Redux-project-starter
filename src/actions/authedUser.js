// Action Types
const LOGIN_AUTHED_USER = 'LOGIN_AUTHED_USER';
const LOGOUT_AUTHED_USER = 'LOGOUT_AUTHED_USER';

// Action Creators
const authenticateUser  = (authedUser) => {
  return {
    type: LOGIN_AUTHED_USER,
    authedUser,
  };
};

const signOutUser  = () => {
  return {
    type: LOGOUT_AUTHED_USER,
  };
};

// Thunks
const authenticateUserWithCredentials = (credential) => {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(
      (user) =>
        user.id === credential.username &&
        user.password === credential.password,
    );
    if (!!user) {
      return dispatch(authenticateUser (user));
    } else {
      return false;
    }
  };
};

//Thunk to handle user sign out.
const signOutCurrentUser  = () => {
  return (dispatch) => {
    dispatch(signOutUser ());
  };
};

export {
  LOGIN_AUTHED_USER,
  LOGOUT_AUTHED_USER,
  authenticateUser ,
  signOutUser ,
  authenticateUserWithCredentials,
  signOutCurrentUser ,
};
