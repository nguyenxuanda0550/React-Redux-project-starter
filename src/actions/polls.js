import { saveQuestion, saveQuestionAnswer } from '../data/api';
import { addUserAnswer, addUserQuestion } from './users';

export const CREATE_NEW_QUESTION = 'CREATE_NEW_QUESTION';
export const ADD_ANSWER_TO_QUESTION = 'ADD_ANSWER_TO_QUESTION';
export const LOAD_QUESTIONS = 'LOAD_QUESTIONS';

const addNewQuestion = (question) => {
  return {
    type: CREATE_NEW_QUESTION,
    question,
  };
};

const addNewAnswerQuestion = (authUser, qid, answer) => {
  return {
    type: ADD_ANSWER_TO_QUESTION,
    authUser,
    qid,
    answer,
  };
};

const receiveQuestions = (questions) => {
  return {
    type: LOAD_QUESTIONS,
    questions,
  };
};

const handleCreateQuestion = (firstOption, secondOption) => {
  return async (dispatch, getState) => {
    const { authUser } = getState();
    const question = {
      optionOneText: firstOption,
      optionTwoText: secondOption,
      author: authUser,
    };
    console.log('question', question)
    try {
      const questionResponse = await saveQuestion(question);
      dispatch(addNewQuestion(questionResponse));
      dispatch(addUserQuestion(questionResponse));
    } catch(error) {
      console.error('Error adding question:', error);
    }   
  };
};

const handleCreateAnswer = (questionId, answer) => {
  return async (dispatch, getState) => {
    const { authUser } = getState();
    const answerObject = {
      authedUser: authUser.id,
      qid: questionId,
      answer,
    };
    const isAdded = await saveQuestionAnswer(answerObject);
    if (isAdded) {
      dispatch(addNewAnswerQuestion(authUser.id, questionId, answer));
      dispatch(addUserAnswer(authUser.id, questionId, answer));
    }
  };
};

export { handleCreateQuestion, handleCreateAnswer, receiveQuestions };
