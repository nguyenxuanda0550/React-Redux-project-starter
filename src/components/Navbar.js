import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { signOutCurrentUser  } from '../actions/authedUser';

const Navbar = ({ dispatch, authUser }) => {
  const logoutHandle = () => {
    dispatch(signOutCurrentUser ());
  };

  return (
    <nav className="navbar navbar-expand-lg navbar-light bg-light">
      <div className="container">
        <Link className="navbar-brand" to={'/'}>
          <span className="fs-4">Employee Poll</span>
        </Link>
        <button
          className="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarNav"
          aria-controls="navbarNav"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span className="navbar-toggler-icon"></span>
        </button>
        <div className="collapse navbar-collapse" id="navbarNav">
          <ul className="navbar-nav me-auto mb-2 mb-lg-0">
            <li className="nav-item">
              <Link className="nav-link" to={'/'}>
                Home
              </Link>
            </li>
            <li className="nav-item">
              <Link className="nav-link" to={'/add'}>
                New Poll
              </Link>
            </li>
            <li className="nav-item">
              <Link className="nav-link" to={'/leaderboard'}>
                Leaderboard
              </Link>
            </li>
          </ul>
          <div className="d-flex align-items-center">
            {authUser ? (
              <>
                <span className="me-3">{authUser.name}</span>
                <button
                  className="btn btn-outline-primary"
                  onClick={logoutHandle}
                >
                  Logout
                </button>
              </>
            ) : (
              <Link
                className="btn btn-primary"
                to={'/login'}
              >
                Login
              </Link>
            )}
          </div>
        </div>
      </div>
    </nav>
  );
};

const mapStateToProps = ({ authUser }) => ({
  authUser,
});

export default connect(mapStateToProps)(Navbar);
