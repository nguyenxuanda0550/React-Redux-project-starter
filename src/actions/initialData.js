import { getInitialData } from '../data/api';
import { receiveUsers } from './users';
import { receiveQuestions } from './polls';

const loadInitialData = () => {
  return async (dispatch) => {
    try{
      const { users, questions } = await getInitialData();
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
    } catch(error) {
      console.error('Error loading initial data:', error);
    }
    
  };
};

export { loadInitialData };
