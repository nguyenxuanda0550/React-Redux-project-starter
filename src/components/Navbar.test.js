import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Navbar from './Navbar';
import { configureStore } from '@reduxjs/toolkit';
import reducer from '../reducers';
import { authenticateUser } from '../actions/authedUser';

// Configure Redux store with the provided reducer
const store = configureStore({ reducer });

describe('Navbar Component', () => {
  // Test to check if the Navbar component renders correctly
  test('should render the component', () => {
    // Dispatch action to set authenticated user
    store.dispatch(
      authenticateUser({
        username: 'tylermcginnis',
        password: '',
      })
    );

    // Render Navbar component wrapped with Redux Provider and React Router
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>
    );

    // Assert that the component is defined and matches the snapshot
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  // Test to check if Navbar displays all elements correctly
  test('should display all elements', () => {
    // Dispatch action to set authenticated user
    store.dispatch(
      authenticateUser({
        username: 'tylermcginnis',
        password: '',
      })
    );

    // Render Navbar component wrapped with Redux Provider and React Router
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Navbar />
        </BrowserRouter>
      </Provider>
    );

    // Select elements by their test IDs
    const homeLinkElement = screen.getByTestId('home-link');
    const newPollLinkElement = screen.getByTestId('new-poll-link');
    const leaderboardLinkElement = screen.getByTestId('leaderboard-link');
    const logoutLinkElement = screen.getByTestId('logout-link');

    // Assert that the text content of the elements is correct
    expect(homeLinkElement.textContent).toBe('Home');
    expect(newPollLinkElement.textContent).toBe('New Poll');
    expect(leaderboardLinkElement.textContent).toBe('Leaderboard');
    expect(logoutLinkElement.textContent).toBe('Logout');
  });
});
