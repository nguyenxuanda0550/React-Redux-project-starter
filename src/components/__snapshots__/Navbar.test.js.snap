// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="navbar navbar-expand-lg navbar-light bg-light"
      >
        <div
          class="container"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            <span
              class="fs-4"
            >
              Employee Poll
            </span>
          </a>
          <button
            aria-controls="navbarNav"
            aria-expanded="false"
            aria-label="Toggle navigation"
            class="navbar-toggler"
            data-bs-target="#navbarNav"
            data-bs-toggle="collapse"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="collapse navbar-collapse"
            id="navbarNav"
          >
            <ul
              class="navbar-nav me-auto mb-2 mb-lg-0"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  data-testid="home-link"
                  href="/"
                >
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  data-testid="new-poll-link"
                  href="/add"
                >
                  New Poll
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link"
                  data-testid="leaderboard-link"
                  href="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
            </ul>
            <div
              class="d-flex align-items-center"
            >
              <span
                class="me-3"
              />
              <button
                class="btn btn-outline-primary"
                data-testid="logout-link"
              >
                Logout
              </button>
            </div>
          </div>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="navbar navbar-expand-lg navbar-light bg-light"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          <span
            class="fs-4"
          >
            Employee Poll
          </span>
        </a>
        <button
          aria-controls="navbarNav"
          aria-expanded="false"
          aria-label="Toggle navigation"
          class="navbar-toggler"
          data-bs-target="#navbarNav"
          data-bs-toggle="collapse"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="collapse navbar-collapse"
          id="navbarNav"
        >
          <ul
            class="navbar-nav me-auto mb-2 mb-lg-0"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-testid="home-link"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-testid="new-poll-link"
                href="/add"
              >
                New Poll
              </a>
            </li>
            <li
              class="nav-item"
            >
              <a
                class="nav-link"
                data-testid="leaderboard-link"
                href="/leaderboard"
              >
                Leaderboard
              </a>
            </li>
          </ul>
          <div
            class="d-flex align-items-center"
          >
            <span
              class="me-3"
            />
            <button
              class="btn btn-outline-primary"
              data-testid="logout-link"
            >
              Logout
            </button>
          </div>
        </div>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
